### PR TITLE
feat: Add Zip streaming

### PR DESCRIPTION
**Pull request by Arno500**
#1118

### CODE REVIEW

**Comment by PHLAK** [comment]
Thank you! Sorry, for not responding sooner but I was out of town earlier this week. I will review this soon.

File: composer.json [hunk 1]
```diff
@@ -35,12 +36,13 @@
...
     "require-dev": {
         "beyondcode/expose": "^2.0",
+        "friendsofphp/php-cs-fixer": "!=3.5.0",
         "johnkary/phpunit-speedtrap": "^4.0",
         "phlak/coding-standards": "^2.0",
-        "friendsofphp/php-cs-fixer": "!=3.5.0",
         "phpstan/phpstan": "^1.0",
         "psy/psysh": "^0.11.0",
         "symfony/var-dumper": "^5.0",
+        "timeweb/phpstan-enum": "^3.1",
```
Comment (PHLAK) [inline composer.json:45]: Please add additional packages.

```suggestion
```

File: composer.lock
Comment (PHLAK) [inline composer.lock:1]: It looks like you did a `composer update` here resulting in several packages being updated that aren't relevant to this PR. Instead, revert the changes here and run `composer require `maennchen/zipstream-php`` to only install the `maennchen/zipstream-php` without affecting existing dependencies.

**Review by PHLAK — COMMENTED** [review]
Please keep the PR focused on only the changes necessary for implementing Zip streaming. I see a lot of changes that aren't relevant to this PR. I believe this is due to running a `composer update` which should not have been done. Please revert the changes to `composer.lock` and then run `composer require maennchen/zipstream-php` to only install that package without updating pre-existing dependencies.

File: app/src/Controllers/ZipController.php [hunk 2]
```diff
@@ -34,41 +34,161 @@ public function __invoke(Request $request, Response $response): ResponseInterfac
...
-
-        return $response->withHeader('Content-Type', 'application/zip')
+        $response = $response
+            ->withHeader('Content-Type', 'application/zip')
             ->withHeader('Content-Disposition', sprintf(
                 'attachment; filename="%s.zip"',
                 $this->generateFileName($path)
-            ));
+            ))
+            ->withHeader('X-Accel-Buffering', 'no');
```
Comment (PHLAK) [inline app/src/Controllers/ZipController.php:43]: What is the purpose of the `X-Accel-Buffering` header here?

Comment (Arno500) [reply app/src/Controllers/ZipController.php:43]: This is to prevent Nginx (and I think any FastCGI server) from buffering the response: https://maennchen.dev/ZipStream-PHP/guide/Nginx.html

File: composer.lock
Comment (Arno500) [reply composer.lock:1]: This is what I did previously, but the CI is running a `composer update` before checking the styling. It appears that php-cs-fixer changed a rule somewhere and the whole CI was failing since it was automatically getting the newer version (see results here: https://github.com/DirectoryLister/DirectoryLister/actions/runs/5100840989/jobs/9296322524).

File: composer.json (hunk 1 above)
Comment (Arno500) [reply composer.json:45]: This package is required to allow PHP Stan to properly evaluate the Enums used in ZipStream (they use it for archive types, for instance). So, I can either use this package to make it aware of them, or ignore the lines (but this is not the suitable option in my opinion).

Comment (PHLAK) [reply composer.json:45]: Which enums are we talking about? Enums weren't introduced until PHP 8.1 and as discussed in the issue I'd like to keep support for PHP 7.4 for now.

File: composer.lock
Comment (PHLAK) [reply composer.lock:1]: > CI is running a `composer update` before checking the styling

That's my bad, it should not be doing an update. My original comment still applies but I'll have to fix CI.

File: composer.json (hunk 1 above)
Comment (Arno500) [reply composer.json:45]: This is a custom Enum module compatible with any version of PHP called php-enum. The ZipStream library uses them here: https://github.com/maennchen/ZipStream-PHP/blob/v2/src/Option/Method.php
Since this is dynamically evaluated, PHPStan is all lost when seeing "unexisting methods" being called.
In the latest version (v3), since they killed the compat with older versions of PHP, they moved to native enums: https://github.com/maennchen/ZipStream-PHP/blob/main/src/CompressionMethod.php

I am making a usage of those enums in the code to indicate the type of compression to apply here: https://github.com/DirectoryLister/DirectoryLister/blob/5b61653b889cec276bdc6bc0cd45a66e6aafdac9/app/src/Controllers/ZipController.php#L62

File: composer.lock
Comment (Arno500) [reply composer.lock:1]: Alright, I will just revert the last commit in this case

**Comment by Arno500** [comment]
Hello @PHLAK !

I updated the review rebased on the master with the CS fixes. I ran the testing suite in my repo and everything seems OK.
There is still an issue with CS, but this issue is also present on master in the latest commit since the bumps and fixes. I'll leave that to you!

**Comment by PHLAK** [comment]
Thanks @Arno500! I have some time off coming up and was planning on getting to this then.

File: composer.json [hunk 3]
```diff
@@ -19,6 +19,7 @@
         "ext-zip": "*",
         "erusev/parsedown-extra": "^0.8.1",
         "filp/whoops": "^2.7",
+        "maennchen/zipstream-php": "^2.4",
```
Comment (PHLAK) [inline composer.json:22]: Now that PHP versions <= 8.1 are [end of life](https://www.php.net/supported-versions.php) we can bump this to v3.

```suggestion
        "maennchen/zipstream-php": "^3.1",
```

Comment (Arno500) [reply composer.json:22]: Alright. It will need refactoring then because they integrated the size calculation beforehand inside the library. So I will remove quite a lot of code and simplify the whole process :)

**Comment by PHLAK** [comment]
I've pushed some dependency updates and fixed CI. Please rebase and push the updates discussed when you have a moment and we can hopefully get this merged.

**Comment by Arno500** [comment]
Done!

File: app/src/Controllers/ZipController.php [hunk 4]
```diff
@@ -33,41 +36,76 @@ public function __invoke(Request $request, Response $response): ResponseInterfac
...
-        foreach ($this->finder->in($path)->files() as $file) {
-            $zip->addFile((string) $file->getRealPath(), $this->stripPath($file, $path));
+        foreach ($files as $file) {
+            $modifiedTime = null;
+
+            try {
+                $modifiedTime = new \DateTime('@' . (int) $file->getMTime());
+            } catch (\Exception $e) {
+                // We couldn't get the creation time, so we don't set it
+            }
```
Comment (PHLAK) [inline app/src/Controllers/ZipController.php:82]: We can use the same approach used in `ModifiedTime.php` here and use `lstat()` as a fall back.

https://github.com/DirectoryLister/DirectoryLister/blob/10795b101167044331e66ddae739d640f99c4a7d/app/src/ViewFunctions/ModifiedTime.php#L20-L24

```suggestion
            try {
                $modifiedTime = new \DateTime('@' . (int) $file->getMTime());
            } catch (RuntimeException $e) {
                $modifiedTime = lstat($file->getPathname())['mtime'];
            }
```

File: app/src/Controllers/ZipController.php [hunk 5]
```diff
@@ -80,4 +118,14 @@ protected function generateFileName(string $path): string
...
     }
+
+    /** Filter a file name to remove invalid characters inside a Zip */
+    protected function filterFilename(string $filename): string
+    {
+        // strip leading slashes from file name
+        // (fixes bug in windows archive viewer)
+        $filename = (string) preg_replace('/^\\/+/', '', $filename);
+
+        return str_replace(['\\', ':', '*', '?', '"', '<', '>', '|'], '_', $filename);
```
Comment (PHLAK) [inline app/src/Controllers/ZipController.php:129]: I may be mistaken but I don't see this method used anywhere.

```suggestion
```

**Review by PHLAK — COMMENTED** [review]
Just a couple of small comments but otherwise this is looking good.

**Comment by PHLAK** [comment]
Thank you for all the hard work @Arno500!
